core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -199,14 +242,20 @@ private[spark] class BarrierCoordinator(
   }
 
   override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
-    case request @ RequestToSync(numTasks, stageId, stageAttemptId, _, _) =>
+    case request @ BarrierRequestToSync(numTasks, stageId, stageAttemptId, _, _, _, _) =>
       // Get or init the ContextBarrierState correspond to the stage attempt.
       val barrierId = ContextBarrierId(stageId, stageAttemptId)
       states.computeIfAbsent(barrierId,
         (key: ContextBarrierId) => new ContextBarrierState(key, numTasks))
       val barrierState = states.get(barrierId)
-
-      barrierState.handleRequest(context, request)
+      barrierState.handleRequest(context, request, "")
+    case request @ AllGatherRequestToSync(numTasks, stageId, stageAttemptId, _, _, _, _, msg) =>
+      // Get or init the ContextBarrierState correspond to the stage attempt.
+      val barrierId = ContextBarrierId(stageId, stageAttemptId)
+      states.computeIfAbsent(barrierId,
+        (key: ContextBarrierId) => new ContextBarrierState(key, numTasks))
+      val barrierState = states.get(barrierId)
+      barrierState.handleRequest(context, request, msg)
```

Comment:
I think we could merge this two cases into one cases and only distinguish the message by:

```
val message = request match {
   case a @ AllGatherRequestToSync => a.msg
   case _ => ""
}
```

Reply:
I'm new to Scala, do you have any idea how to get around the fact that `case request @ A | B` is illegal in Scala? (IE alternative matching doesnt support variables)